public/js/components/App.js

```diff
@@ -12,30 +14,106 @@ const SplitBox = createFactory(require("./SplitBox"));
 const RightSidebar = createFactory(require("./RightSidebar"));
 const SourceTabs = createFactory(require("./SourceTabs"));
 const SourceFooter = createFactory(require("./SourceFooter"));
-const { getSources, getBreakpoints } = require("../selectors");
+const Autocomplete = createFactory(require("./Autocomplete"));
+const { getSelectedSource, getSources, getBreakpoints } = require("../selectors");
+const { endTruncateStr } = require("../util/utils");
 
 const App = React.createClass({
   propTypes: {
     sources: PropTypes.object,
     selectedSource: PropTypes.object,
+    selectSource: PropTypes.func,
     breakpoints: PropTypes.object
   },
 
   displayName: "App",
 
+  getInitialState() {
+    return {
+      searchOn: false
+    };
+  },
+
+  componentDidMount() {
+    window.addEventListener("keydown", this.toggleSourcesSearch, false);
+  },
+
+  componentWillUnmount() {
+    window.removeEventListener("keydown", this.toggleSourcesSearch, false);
+  },
+
+  toggleSourcesSearch(e) {
+    if (e.metaKey && e.key == "p") {
+      this.setState({ searchOn: !this.state.searchOn });
+      e.preventDefault();
+    }
+  },
+
+  renderSourcesSearch() {
+    function getSourcePath(source) {
+      const url = source.get("url") || "";
+      const path = (new URL(url)).pathname;
+      return endTruncateStr(path, 50);
+    }
+
+    function searchResults(sources) {
+      return sources.valueSeq()
+        .filter(source => !!source.get("url"))
+        .map(source => ({
+          value: getSourcePath(source),
+          title: getSourcePath(source).split("/").pop(),
+          subtitle: getSourcePath(source),
+          id: source.get("id")
+        }))
+        .toJS();
+    }
+
+    return Autocomplete({
+      selectItem: (result) => {
+        this.props.selectSource(result.id);
+        this.setState({ searchOn: false });
+      },
+      items: searchResults(this.props.sources)
+    });
+  },
+
+  renderEditor() {
+    return dom.div(
+      { className: "editor-container" },
+      SourceTabs(),
+      Editor(),
+      SourceFooter()
+    );
+  },
+
+  renderWelcomeBox() {
+    return dom.div(
+      { className: "welcomebox" },
+      "Hit Cmd + P to open a file"
+    );
+  },
+
+  renderCenterPane() {
+    let centerPane;
+    if (this.state.searchOn) {
```

Comment:
This kind of stuff, because it's very basic (and important) app state, might be nice to store in redux. But we can easily move it out later.